src/GraphQl/Resolver/Factory/CollectionResolverFactory.php

```diff
@@ -87,17 +87,20 @@ public function __invoke(?string $resourceClass = null, ?string $rootClass = nul
                 $collection = $queryResolver($collection, $resolverContext);
             }
 
-            ($this->securityStage)($resourceClass, $operationName, $resolverContext + [
-                'extra_variables' => [
-                    'object' => $collection,
-                ],
-            ]);
-            ($this->securityPostDenormalizeStage)($resourceClass, $operationName, $resolverContext + [
-                'extra_variables' => [
-                    'object' => $collection,
-                    'previous_object' => $this->clone($collection),
-                ],
-            ]);
+            // Only perform security stage on the top-level query
+            if (null === $source) {
```

Comment:
I think it's dangerous to do this. If the developer has added security in the related resource, they think it will be always applied in all contexts, don't you think?
Only applying the security at the top level means that if a user has access to a resource, it can access to secured resources by traversing the relations (if the relations have not been secured with your change of course).
I think the better approach is to change the operation name to be always `collection_query`.

Reply:
I do see your point, and would certainly want to update the documentation to make it clearer if we do go this route.
That said, I see the `ApiResource` GraphQL `item_query`, `collection_query`, etc. as definining the actual GraphQL entrypoints i.e. the top-level queries.
Association collections are different collections to the main resource collection - they're filtered. The `collection_query` in `ApiResource` exposes the entire collection. I don't think it makes sense to use the same security on associations, as the requirements may be different. In my example (in https://github.com/api-platform/core/issues/4169) you would want admins to see the whole collection of documents, and not let normal users see them all, instead letting users only see their own documents via the `user.documents` association. Using the `collection_query` security on the association would mean only admins could ever see it.
You may also not want to have a `collection_query` exposing the whole collection as an endpoint, but still want to have associations. Relying on there being a `collection_query` present wouldn't work in this case.
Unless I've found a bug, in 2.6.3 at least, this PR makes collection security work the same way as single associations already do. `item_query` security from the association resource is not run on associations - meaning they are already exposed with no way to protect them. This is covered by the following two tests I've added:
- https://github.com/api-platform/core/pull/4143/files#diff-14861a046eb66cb65d4007864a6fbe8d08b64a6cd05a61f21d5cab03b7eb7525R193 - shows that it's not possible for a user to access an `item_query` secured resource directly
- https://github.com/api-platform/core/pull/4143/files#diff-14861a046eb66cb65d4007864a6fbe8d08b64a6cd05a61f21d5cab03b7eb7525R254 - shows that it's possible to access the `item_query` secured resource via a association. The `item_query` security is never run.

I hope that all makes sense!